apps/dashboard/src/hooks/use-form-autosave.ts

```diff
@@ -1,24 +1,81 @@
 // useFormAutosave.ts
 import { useEffect } from 'react';
 import { UseFormReturn, useWatch, FieldValues } from 'react-hook-form';
+import { useDataRef } from './use-data-ref';
+import { useDebounce } from './use-debounce';
 
-export function useFormAutosave<T extends FieldValues>(form: UseFormReturn<T>, save: (data: T) => void): void {
+const TEN_SECONDS = 10 * 1000;
+
+export function useFormAutosave<U extends Record<string, unknown>, T extends FieldValues = FieldValues>({
+  previousData,
+  form: propsForm,
+  isReadOnly,
+  update,
+}: {
+  previousData: U;
+  form: UseFormReturn<T>;
+  isReadOnly?: boolean;
+  update: (data: U) => void;
+}) {
+  const formRef = useDataRef(propsForm);
   const watchedData = useWatch<T>({
-    control: form.control,
+    control: propsForm.control,
   });
 
+  const onUpdate = async (data: T) => {
+    // use the form reference instead of descructuring the props to avoid stale closures
+    const form = formRef.current;
+    const isDirty = form.formState.isDirty;
+    if (!isDirty || isReadOnly) {
+      return;
+    }
+    const isValid = await form.trigger();
+    if (!isValid) {
+      return;
+    }
+
+    const values = { ...previousData, ...data };
+    // reset the dirty fields right away because on slow networks the patch request might take a while
+    // so other blur/change events might trigger in the meantime
+    form.reset(values);
```

Comment:
Doesn't this reset the entire form?

Reply:
yes, but we don't care actually because it's updated with the latest form values, and also because of the above comment

Reply:
The only issue I can think of with resetting the form with its current values is that in case of the response being different (a field changes upon submission by the api or an error is returned for it), we're left with stale data.

Reply:
talked about that in DM